# Vindiata Consulting Private Limited Internship Assignment


### Case Study

### Part - A

###### 	Airline A is currently flying a large international network. Their flight operations are shown in the tab "Operations". Various aircraft characteristics are listed in tab "AC characteristics". They would like to know the following:	


* 25 points	1)	What is the total cost by aircraft type for the year? (Suggested tools: Any of Python/SQL )														
		Note:	Total cost = Hours flown * Costs/hour
        
        
        
* 25 points	2)	Which aircraft type  has the lowest cost per seat per km flown?														


### Solution : 

#### IMPORTING DATA FROM THE GIVEN EXCEL FILE

* Evaluated this Business Problem via python's most readible and functional library i.e. Pandas


* Data is copied from "AC Characterstics" sheet  and pasted to "formated_AC_characterstics" sheet to make it more readable in  pandas format in order to get rid off irrelavant enteries.


* Data is copied from "Operations" sheet  and pasted to "formated_Operations" sheet to make it more readable in  pandas format in order to get rid off irrelavant enteries.


* Data is copied from "City_pairs" sheet  and pasted to "formated_City_paird" sheet to make it more readable in  pandas format in order to get rid off irrelavant enteries.


**Ac_characterstics sheet:**

In [48]:
import pandas as pd 
df = pd.read_excel('Data Science Case study Vindiata.xls',sheet_name = 'formated_AC_characterstics')
df

Aircraft Type  Range (Km)  Ave. Speed (km/h)  Number of Seats  \
0          A320        5000                800              150   
1          A330        8000                900              250   
2          B737        5000                800              150   
3          B747       10000                900              350   
4          Q400        1500                750               90   
5         ATR72        1000                650               75   

   Costs per flight hour  
0                   5000  
1                   7500  
2                   5100  
3                  12000  
4                   3500  
5                   2750

**Operations_sheet**

In [49]:
opr_sheet = pd.read_excel('Data Science Case study Vindiata.xls',sheet_name = 'formatted_Operations')
col_names = ['Jan-14','Feb-14','Mar-14','Apr-14','May-14','Jun-14','Jul-14','Aug-14','Sep-14','Oct-14','Nov-14','Dec-14']
opr_sheet.rename(columns=dict(zip(opr_sheet.columns[2:], col_names)), inplace=True)
opr_sheet

Aircraft Name Aircraft Type  Jan-14  Feb-14  Mar-14  Apr-14  May-14  \
0             A           A320     250     251     248     253     243   
1              B          A330     300     301     298     303     293   
2              C          B737     245     246     243     248     238   
3              D          B747     310     311     308     313     303   
4              E          Q400     200     201     198     203     193   
5              F         ATR72     195     196     193     198     188   
6              G          B747     300     301     298     303     293   
7              H          Q400     195     196     193     198     188   
8             I          ATR72     200     201     198     203     193   
9              J          A330     310     311     308     313     303   
10             K          B737     250     251     248     253     243   
11             L          A320     245     246     243     248     238   
12             M          A330     300     301     298     303     293   
13             N          B737     245     246     243     248     238   
14             O          A320     250     251     248     253     243   
15             P          A320     245     246     243     248     238   
16             Q          Q400     195     196     193     198     188   
17             R         ATR72     200     201     198     203     193   
18             S          A330     310     311     308     313     303   
19             T          A330     300     301     298     303     293   
20             U          B747     310     311     308     313     303   
21             V          Q400     200     201     198     203     193   
22             W         ATR72     200     201     198     203     193   
23             X          A320     250     251     248     253     243   
24             Y          A330     300     301     298     303     293   
25             Z          B737     245     246     243     248     238   

    Jun-14  Jul-14  Aug-14  Sep-14  Oct-14  Nov-14  Dec-14  
0      245     248     246     256     256     251     252  
1      295     298     296     306     306     301     302  
2      240     243     241     251     251     246     247  
3      305     308     306     316     316     311     312  
4      195     198     196     206     206     201     202  
5      190     193     191     201     201     196     197  
6      295     298     296     306     306     301     302  
7      190     193     191     201     201     196     197  
8      195     198     196     206     206     201     202  
9      305     308     306     316     316     311     312  
10     245     248     246     256     256     251     252  
11     240     243     241     251     251     246     247  
12     295     298     296     306     306     301     302  
13     240     243     241     251     251     246     247  
14     245     248     246     256     256     251     252  
15     240     243     241     251     251     246     247  
16     190     193     191     201     201     196     197  
17     195     198     196     206     206     201     202  
18     305     308     306     316     316     311     312  
19     295     298     296     306     306     301     302  
20     305     308     306     316     316     311     312  
21     195     198     196     206     206     201     202  
22     195     198     196     206     206     201     202  
23     245     248     246     256     256     251     252  
24     295     298     296     306     306     301     302  
25     240     243     241     251     251     246     247

**City_pairs_sheet**

In [50]:
cp_sheet = pd.read_excel('Data Science Case study Vindiata.xls',sheet_name = 'formatted_city_pairs')
cp_sheet.rename(columns = {'Pass. Demand \n(per day)':'Pass. Demand (per day)'}, inplace = True)
cp_sheet

Origin City Desitnation City  Pass. Demand (per day)  Distance (km)
0          AA               BB                     420           3000
1          BB               CC                     450           6500
2          CC               AA                     300            400
3          AA               DD                     300           1000

In [51]:
#creating a new column "Total_Hr_Flown" in "opr_sheet":

opr_sheet['Total_Hr_Flown'] = opr_sheet.iloc[:,2:].sum(axis = 1)
opr_sheet.head()


Aircraft Name Aircraft Type  Jan-14  Feb-14  Mar-14  Apr-14  May-14  Jun-14  \
0            A           A320     250     251     248     253     243     245   
1             B          A330     300     301     298     303     293     295   
2             C          B737     245     246     243     248     238     240   
3             D          B747     310     311     308     313     303     305   
4             E          Q400     200     201     198     203     193     195   

   Jul-14  Aug-14  Sep-14  Oct-14  Nov-14  Dec-14  Total_Hr_Flown  
0     248     246     256     256     251     252            2999  
1     298     296     306     306     301     302            3599  
2     243     241     251     251     246     247            2939  
3     308     306     316     316     311     312            3719  
4     198     196     206     206     201     202            2399

In [52]:
#creating a copy of dataframe "opr_sheet" so that the changes won't reflect to the original dataframe

opr_sheet_new = opr_sheet.copy()
opr_sheet_new = opr_sheet_new.groupby(['Aircraft Type']).sum()

#merging 2 dataframes i.e. "Ac_characterstics" and "Operations"

new_df = df.merge(opr_sheet_new, how = 'left', on = 'Aircraft Type')
new_df

Aircraft Type  Range (Km)  Ave. Speed (km/h)  Number of Seats  \
0          A320        5000                800              150   
1          A330        8000                900              250   
2          B737        5000                800              150   
3          B747       10000                900              350   
4          Q400        1500                750               90   
5         ATR72        1000                650               75   

   Costs per flight hour  Jan-14  Feb-14  Mar-14  Apr-14  May-14  Jun-14  \
0                   5000    1240    1245    1230    1255    1205    1215   
1                   7500    1820    1826    1808    1838    1778    1790   
2                   5100     985     989     977     997     957     965   
3                  12000     920     923     914     929     899     905   
4                   3500     790     794     782     802     762     770   
5                   2750     795     799     787     807     767     775   

   Jul-14  Aug-14  Sep-14  Oct-14  Nov-14  Dec-14  Total_Hr_Flown  
0    1230    1220    1270    1270    1245    1250           14875  
1    1808    1796    1856    1856    1826    1832           21834  
2     977     969    1009    1009     989     993           11816  
3     914     908     938     938     923     926           11037  
4     782     774     814     814     794     798            9476  
5     787     779     819     819     799     803            9536

### 1) Total Cost As Per Aircraft Type

#### Total cost = Hours flown * Costs / hour


In [53]:
new_df['Total_Cost(USD)'] = new_df['Total_Hr_Flown'] * new_df['Costs per flight hour']
new_df[['Aircraft Type','Total_Cost(USD)' ]]

Aircraft Type  Total_Cost(USD)
0          A320         74375000
1          A330        163755000
2          B737         60261600
3          B747        132444000
4          Q400         33166000
5         ATR72         26224000

### 2) Cost per seat per Km by Aircraft Type :


**Cost per seat per km = Total cost(USD) / (No. of seats x Distance Covered in (km) )  ------------------------ 1)**

**Distance Covered in (km) : Total hours x Avg. speed   ----------------------------------2)**



* Substituting equation 2) in equation 1):

* New Formula Derieved: 

**Cost per seat per km = Total cost(USD) / (No. of seats x Total hours x Avg. speed)**





In [54]:
new_df['Cost per seat per km'] = new_df['Total_Cost(USD)'] / (new_df['Number of Seats'] * new_df['Total_Hr_Flown'] * new_df['Ave. Speed (km/h)'])

## sorting the "Cost per seat per km" column in ascending order :
new_df[['Aircraft Type', 'Cost per seat per km']].sort_values(by = ['Cost per seat per km'])

Aircraft Type  Cost per seat per km
1          A330              0.033333
3          B747              0.038095
0          A320              0.041667
2          B737              0.042500
4          Q400              0.051852
5         ATR72              0.056410

**Conclusion** : Aircraft  **A330** is the best suited aircraft type as it the most cost effective aircraft which costs **0.033333 USD** only.

### Part - B

##### Airline B is a startup airline and is considering flying on the city pairs as shown in tab "City Pairs". They would like to know:

**1.) Which aircraft types are best suited for their operation?  (Suggested tools: Any of Python/SQL )**

**Note:**

* This is an optimization between 3 factors: Range, Passenger demand, costs. 

* Hence the aircraft which is capable of flying between 2 cities (range), has enough seats to meet the demand, and has the lowest cost will win. 
    
* Please also note that the passenger demand is per day, hence an aircraft can make several trips between 2 cities to meet that demand.		
    
* 1 aircraft type will always be most suitable for a given city pair. Hence resultant fleet would be a mix of different aircraft types.	

* In case you are unable to get to the final solution, please lay out the step-by-step approach to be used.										


### Solution :

In [55]:
new_df

Aircraft Type  Range (Km)  Ave. Speed (km/h)  Number of Seats  \
0          A320        5000                800              150   
1          A330        8000                900              250   
2          B737        5000                800              150   
3          B747       10000                900              350   
4          Q400        1500                750               90   
5         ATR72        1000                650               75   

   Costs per flight hour  Jan-14  Feb-14  Mar-14  Apr-14  May-14  Jun-14  \
0                   5000    1240    1245    1230    1255    1205    1215   
1                   7500    1820    1826    1808    1838    1778    1790   
2                   5100     985     989     977     997     957     965   
3                  12000     920     923     914     929     899     905   
4                   3500     790     794     782     802     762     770   
5                   2750     795     799     787     807     767     775   

   Jul-14  Aug-14  Sep-14  Oct-14  Nov-14  Dec-14  Total_Hr_Flown  \
0    1230    1220    1270    1270    1245    1250           14875   
1    1808    1796    1856    1856    1826    1832           21834   
2     977     969    1009    1009     989     993           11816   
3     914     908     938     938     923     926           11037   
4     782     774     814     814     794     798            9476   
5     787     779     819     819     799     803            9536   

   Total_Cost(USD)  Cost per seat per km  
0         74375000              0.041667  
1        163755000              0.033333  
2         60261600              0.042500  
3        132444000              0.038095  
4         33166000              0.051852  
5         26224000              0.056410

In [56]:
## extracting relevant columns from new_df dataframe for further analysis:

new_df.drop(new_df.iloc[:, 5:17], axis = 1, inplace = True)
new_df.drop(new_df.columns[[2,4,5,6]], axis = 1, inplace = True)
new_df

Aircraft Type  Range (Km)  Number of Seats  Cost per seat per km
0          A320        5000              150              0.041667
1          A330        8000              250              0.033333
2          B737        5000              150              0.042500
3          B747       10000              350              0.038095
4          Q400        1500               90              0.051852
5         ATR72        1000               75              0.056410

#### Examining relevant airlines for traveling between the source and target destination

**Procedure for same aircraft type between a city pair:**

**Exploring, if the range of the aircraft is greater than the distance between the source and target destination:**

   **If the condition is met, then we will be calculating the cost of the trip using the following formula:**
   
   * Total cost of the trip = Cost per seat per km x Number of Seats x Distance(km) x Number of trips 
   

* Here, Number of trips = Pass. Demand per day / Number of Seats

In [57]:
cp_sheet

Origin City Desitnation City  Pass. Demand (per day)  Distance (km)
0          AA               BB                     420           3000
1          BB               CC                     450           6500
2          CC               AA                     300            400
3          AA               DD                     300           1000

In [58]:
new_df

Aircraft Type  Range (Km)  Number of Seats  Cost per seat per km
0          A320        5000              150              0.041667
1          A330        8000              250              0.033333
2          B737        5000              150              0.042500
3          B747       10000              350              0.038095
4          Q400        1500               90              0.051852
5         ATR72        1000               75              0.056410

In [59]:
import numpy as np
for a,b,c,d in cp_sheet.itertuples(index=False):
    new_df[str(a)+'-'+str(b)+' cost'] = np.where(new_df['Range (Km)'] >= d, (new_df['Cost per seat per km'] * new_df['Number of Seats'] * d * np.ceil(c / df['Number of Seats'])).astype(int), "NaN")

# Describing dataframe with added travel cost
new_df

Aircraft Type  Range (Km)  Number of Seats  Cost per seat per km AA-BB cost  \
0          A320        5000              150              0.041667      56250   
1          A330        8000              250              0.033333      50000   
2          B737        5000              150              0.042500      57375   
3          B747       10000              350              0.038095      80000   
4          Q400        1500               90              0.051852        NaN   
5         ATR72        1000               75              0.056410        NaN   

  BB-CC cost CC-AA cost AA-DD cost  
0        NaN       5000      12500  
1     108333       6666      16666  
2        NaN       5100      12750  
3     173333       5333      13333  
4        NaN       7466      18666  
5        NaN       6769      16923

The following table summarizes the optimal aircraft type for each city pair along with the number of trips required:

In [60]:

index = ['AA-BB','BB-CC','CC-AA','AA-DD'] 
column = ['Optimal aircraft type','Number of trips']

# No. of trips = Pass. demand per day / No. of seats
curr_data = [['A330', np.ceil(420/250)], ['A330', np.ceil(450/250)], ['A320', np.ceil(300/150)], ['A320', np.ceil(300/150)]]
result = pd.DataFrame(curr_data, index = index, columns = column)
result

Optimal aircraft type  Number of trips
AA-BB                  A330              2.0
BB-CC                  A330              2.0
CC-AA                  A320              2.0
AA-DD                  A320              2.0

**Analysis for a mixture of aircraft type between a city pair:**

**We are manually finding the optimal aircraft types for each city pair.**

In [61]:
cp_sheet

Origin City Desitnation City  Pass. Demand (per day)  Distance (km)
0          AA               BB                     420           3000
1          BB               CC                     450           6500
2          CC               AA                     300            400
3          AA               DD                     300           1000

In [62]:
new_df

Aircraft Type  Range (Km)  Number of Seats  Cost per seat per km AA-BB cost  \
0          A320        5000              150              0.041667      56250   
1          A330        8000              250              0.033333      50000   
2          B737        5000              150              0.042500      57375   
3          B747       10000              350              0.038095      80000   
4          Q400        1500               90              0.051852        NaN   
5         ATR72        1000               75              0.056410        NaN   

  BB-CC cost CC-AA cost AA-DD cost  
0        NaN       5000      12500  
1     108333       6666      16666  
2        NaN       5100      12750  
3     173333       5333      13333  
4        NaN       7466      18666  
5        NaN       6769      16923

**City Pair AA-BB**

In [63]:
df1 = new_df.iloc[:,:4].copy()

# Travel cost and remaining passengers after 1st flight
df1['AA-BB cost 1'] = np.where(df1['Range (Km)'] >= 3000, (df1['Cost per seat per km'] * df1['Number of Seats'] * 3000).astype(int), 0)
df1['AA-BB remain 1'] = np.where(np.logical_and(420 - df1['Number of Seats'] >= 0, df1['AA-BB cost 1'] != 0), 420 - df1['Number of Seats'], 0)

# Total travel cost and remaining passengers after 2nd flight
df1['AA-BB cost 2'] = np.where(df1['Range (Km)'] >= 3000, (df1['Cost per seat per km'] * df1['Number of Seats'] * 3000).astype(int) + df1['AA-BB cost 1'], 0)
df1['AA-BB remain 2'] = np.where((df1['AA-BB remain 1'] - df1['Number of Seats']) >= 0, df1['AA-BB remain 1'] - df1['Number of Seats'], 0)
df1

Aircraft Type  Range (Km)  Number of Seats  Cost per seat per km  \
0          A320        5000              150              0.041667   
1          A330        8000              250              0.033333   
2          B737        5000              150              0.042500   
3          B747       10000              350              0.038095   
4          Q400        1500               90              0.051852   
5         ATR72        1000               75              0.056410   

   AA-BB cost 1  AA-BB remain 1  AA-BB cost 2  AA-BB remain 2  
0         18750             270         37500             120  
1         25000             170         50000               0  
2         19125             270         38250             120  
3         40000              70         80000               0  
4             0               0             0               0  
5             0               0             0               0

**Conclusion for AA-BB: 2 aircrafts of type A330 are preferable.**



**City Pair BB-CC**



In [64]:
df2 = new_df.iloc[:,:4].copy()

# Travel cost and remaining passengers after 1st flight
df2['BB-CC cost 1'] = np.where(df2['Range (Km)'] >= 6500, (df2['Cost per seat per km'] * df2['Number of Seats'] * 6500).astype(int), 0)
df2['BB-CC remain 1'] = np.where(np.logical_and(450 - df2['Number of Seats'] >= 0, df2['BB-CC cost 1'] != 0), 450 - df2['Number of Seats'], 0)

# Total travel cost and remaining passengers after 2nd flight
df2['BB-CC cost 2'] = np.where(df2['Range (Km)'] >= 6500, (df2['Cost per seat per km'] * df2['Number of Seats'] * 6500).astype(int) + df2['BB-CC cost 1'], 0)
df2['BB-CC remain 2'] = np.where((df2['BB-CC remain 1'] - df2['Number of Seats']) >= 0, df2['BB-CC remain 1'] - df2['Number of Seats'], 0)
df2

Aircraft Type  Range (Km)  Number of Seats  Cost per seat per km  \
0          A320        5000              150              0.041667   
1          A330        8000              250              0.033333   
2          B737        5000              150              0.042500   
3          B747       10000              350              0.038095   
4          Q400        1500               90              0.051852   
5         ATR72        1000               75              0.056410   

   BB-CC cost 1  BB-CC remain 1  BB-CC cost 2  BB-CC remain 2  
0             0               0             0               0  
1         54166             200        108332               0  
2             0               0             0               0  
3         86666             100        173332               0  
4             0               0             0               0  
5             0               0             0               0

**Conclusion for BB-CC: As most of the aircrafts are not compatible for this journey, 2 aircrafts of type A330 are preferable.**

**City Pair CC-AA**

In [65]:
df3 = new_df.iloc[:,:4].copy()

# Travel cost and remaining passengers after 1st flight
df3['CC-AA cost 1'] = np.where(df3['Range (Km)'] >= 400, (df3['Cost per seat per km'] * df3['Number of Seats'] * 400).astype(int), 0)
df3['CC-AA remain 1'] = np.where(np.logical_and(300 - df3['Number of Seats'] >= 0, df3['CC-AA cost 1'] != 0), 300 - df3['Number of Seats'], 0)

# Total travel cost and remaining passengers after 2nd flight
df3['CC-AA cost 2'] = np.where(np.logical_and(df3['Range (Km)'] >= 400, df3['CC-AA remain 1'] != 0), (df3['Cost per seat per km'] * df3['Number of Seats'] * 400).astype(int) + df3['CC-AA cost 1'], df3['CC-AA cost 1'])
df3['CC-AA remain 2'] = np.where((df3['CC-AA remain 1'] - df3['Number of Seats']) >= 0, df3['CC-AA remain 1'] - df3['Number of Seats'], 0)
df3

Aircraft Type  Range (Km)  Number of Seats  Cost per seat per km  \
0          A320        5000              150              0.041667   
1          A330        8000              250              0.033333   
2          B737        5000              150              0.042500   
3          B747       10000              350              0.038095   
4          Q400        1500               90              0.051852   
5         ATR72        1000               75              0.056410   

   CC-AA cost 1  CC-AA remain 1  CC-AA cost 2  CC-AA remain 2  
0          2500             150          5000               0  
1          3333              50          6666               0  
2          2550             150          5100               0  
3          5333               0          5333               0  
4          1866             210          3732             120  
5          1692             225          3384             150

**Conclusion for CC-AA: If 1 A330 and 1 ATR72 are used, the cost is still 25 USD more than using 2 aircrafts of type A320. Hence, A320 is preferred over A330 & ATR72**

**City Pair AA-DD**

In [66]:
df4 = new_df.iloc[:,:4].copy()

# Travel cost and remaining passengers after 1st flight
df4['AA-DD cost 1'] = np.where(df4['Range (Km)'] >= 1000, (df4['Cost per seat per km'] * df4['Number of Seats'] * 1000).astype(int), 0)
df4['AA-DD remain 1'] = np.where(np.logical_and(300 - df4['Number of Seats'] >= 0, df4['AA-DD cost 1'] != 0), 300 - df4['Number of Seats'], 0)

# Total travel cost and remaining passengers after 2nd flight
df4['AA-DD cost 2'] = np.where(np.logical_and(df4['Range (Km)'] >= 1000, df4['AA-DD remain 1'] != 0), (df4['Cost per seat per km'] * df4['Number of Seats'] * 1000).astype(int) + df4['AA-DD cost 1'], df4['AA-DD cost 1'])
df4['AA-DD remain 2'] = np.where((df4['AA-DD remain 1'] - df4['Number of Seats']) >= 0, df4['AA-DD remain 1'] - df4['Number of Seats'], 0)
df4

Aircraft Type  Range (Km)  Number of Seats  Cost per seat per km  \
0          A320        5000              150              0.041667   
1          A330        8000              250              0.033333   
2          B737        5000              150              0.042500   
3          B747       10000              350              0.038095   
4          Q400        1500               90              0.051852   
5         ATR72        1000               75              0.056410   

   AA-DD cost 1  AA-DD remain 1  AA-DD cost 2  AA-DD remain 2  
0          6250             150         12500               0  
1          8333              50         16666               0  
2          6375             150         12750               0  
3         13333               0         13333               0  
4          4666             210          9332             120  
5          4230             225          8460             150

**Conclusion for AA-DD: In extension to the previous scenerio, 2 aircrafts of type "A320" will be apt than a mixture of disntict types.**

### Based on the concrete findings for optimal Aircrafts, the following Infrences are made after considering all the required parameters:
    
* **4** Aircrafts of **A330** and **4** Aircrafts of **A320** are the most optimal options to travel for **Airline B**